File: inventory-allocator/src/InventoryAllocatorTest.py
import unittest
from InventoryAllocator import InventoryAllocator

##########################################
#        INSTRUCTIONS FOR TESTING        #
##########################################
#										 #
# run the following code:				 #
#										 #
# $ python InventoryAllocatorTest.py     #
#										 #
##########################################


class InventoryAllocatorTest(unittest.TestCase):

    def test_happy_case(self):
    	print("\nStarting Test: test_happy_case")
    	IA = InventoryAllocator()
    	order = { 'apple': 5}
    	store_1 = {'name':'owd', 'inventory': { 'apple': 5}}
    	inventory_dist = [store_1]
    	expected = [{'owd': {'apple': 5}}]
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_not_enough_inventory(self):
    	print("\nStarting Test: test_not_enough_inventory")
    	IA = InventoryAllocator()
    	order = { 'apple': 5}
    	store_1 = {'name':'owd', 'inventory': { 'apple': 2 }}
    	inventory_dist = [store_1]
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_should_split(self):
    	print("\nStarting Test: test_should_split")
    	IA = InventoryAllocator()
    	order = { 'apple': 5, 'banana': 10}
    	store_1 = {'name':'owd', 'inventory': { 'apple': 2 , 'banana': 1}}
    	store_2 = {'name':'dsw', 'inventory': { 'apple': 3 , 'banana': 12}}
    	inventory_dist = [store_1, store_2]
    	expected = [{'owd': {'apple': 2, 'banana': 1}}, {'dsw': {'apple':3, 'banana': 9}}]
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_not_at_any_stores(self):
    	print("\nStarting Test: test_not_at_any_stores")
    	IA = InventoryAllocator()
    	order = { 'apple': 5, 'banana': 10}
    	store_1 = {'name':'owd', 'inventory': { 'pickles': 2 , 'grapes': 1}}
    	store_2 = {'name':'dsw', 'inventory': { 'chips': 3 , 'juice': 12}}
    	inventory_dist = [store_1, store_2]
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_zero_inventory_stores(self):
    	print("\nStarting Test: test_zero_inventory_stores")
    	IA = InventoryAllocator()
    	order = { 'apple': 5, 'banana': 10}
    	store_1 = {'name':'owd', 'inventory': { 'apple': 0 , 'banana': 0}}
    	store_2 = {'name':'dsw', 'inventory': { 'chips': 0 , 'banana': 0}}
    	inventory_dist = [store_1, store_2]
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_no_stores(self):
    	print("\nStarting Test: test_no_stores")
    	IA = InventoryAllocator()
    	order = { 'apple': 5, 'banana': 10}
    	inventory_dist = []
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_no_order(self):
    	print("\nStarting Test: test_no_order")
    	IA = InventoryAllocator()
    	order = {}
    	store_1 = {'name':'owd', 'inventory': { 'apple': 0 , 'banana': 0}}
    	store_2 = {'name':'dsw', 'inventory': { 'chips': 0 , 'banana': 0}}
    	inventory_dist = [store_1, store_2]
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")

    def test_no_order_no_stores(self):
    	print("\nStarting Test: test_no_order_no_stores")
    	IA = InventoryAllocator()
    	order = {}
    	inventory_dist = []
    	expected = []
    	self.assertEqual(IA.solution(order, inventory_dist), expected)
    	print("PASSED!")


if __name__ == '__main__':
	print('\n'+'='*40)
	print('= Running Tests for InventoryAllocator =')
	print('='*40,'\n')
	unittest.main()

File: inventory-allocator/src/InventoryAllocatorTest.java
import org.junit.Assert;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

class InventoryAllocatorTest {

    @Test
    void fulfillOrder() {
        Warehouse owd = new Warehouse("owd");
        owd.addToInventory("apple", 5);
        owd.addToInventory("banana", 6);
        owd.addToInventory("peach", 8);
        owd.addToInventory("guava", 5);
        System.out.println("\n----- owd warehouse -----\n");
        System.out.println(owd);

        Warehouse dm = new Warehouse("dm");
        dm.addToInventory("apple", 5);
        dm.addToInventory("peach", 2);
        dm.addToInventory("pineapple", 7);
        dm.addToInventory("banana", 6);
        dm.addToInventory("guava", 5);
        System.out.println("\n----- dm warehouse -----\n");
        System.out.println(dm);

        List<Warehouse> warehouses = new ArrayList<>();
        warehouses.add(owd);
        warehouses.add(dm);

        InventoryAllocator allocator = new InventoryAllocator(warehouses);

        //HAPPY CASE - fulfilled without splitting
        Order HAPPY_ORDER = new Order();
        HAPPY_ORDER.addItem("apple", 5);
        HAPPY_ORDER.addItem("pineapple", 3);
        System.out.println("\n----- HAPPY_ORDER -----\n");
        System.out.println(HAPPY_ORDER);

        Fulfillment HAPPY_CASE = new Fulfillment();
        HAPPY_CASE.addFulfillment("owd", "apple", 5);
        HAPPY_CASE.addFulfillment("dm", "pineapple", 3);
        System.out.println("\n----- HAPPY_CASE -----\n");
        System.out.println(HAPPY_CASE);

        Fulfillment actualHappyCase = allocator.fulfillOrder(HAPPY_ORDER);
        // HAPPY_CASE ASSERTION
        Assert.assertEquals(HAPPY_CASE, actualHappyCase);


        // SAD CASE - nothing fulfilled
        Order SAD_ORDER = new Order();
        SAD_ORDER.addItem("peach", 12);

        System.out.println("\n----- SAD_ORDER -----\n");
        System.out.println(SAD_ORDER);

        Fulfillment SAD_CASE = new Fulfillment();
        System.out.println("\n----- SAD_CASE -----\n");
        System.out.println(SAD_CASE);

        Fulfillment actualSadCase = allocator.fulfillOrder(SAD_ORDER);
        // SAD_CASE ASSERTION
        Assert.assertEquals(SAD_CASE, actualSadCase);


        // SPLIT CASE - fulfilled by splitting / combining
        Order SPLIT_ORDER = new Order();
        SPLIT_ORDER.addItem("banana", 7);
        SPLIT_ORDER.addItem("guava", 10);

        System.out.println("\n----- SPLIT_ORDER -----\n");
        System.out.println(SPLIT_ORDER);

        Fulfillment SPLIT_CASE = new Fulfillment();
        SPLIT_CASE.addFulfillment("owd", "banana", 6);
        SPLIT_CASE.addFulfillment("dm", "banana", 1);
        SPLIT_CASE.addFulfillment("owd", "guava", 5);
        SPLIT_CASE.addFulfillment("dm", "guava", 5);

        System.out.println("\n----- SPLIT_CASE -----\n");
        System.out.println(SPLIT_CASE);

        Fulfillment actualSplitCase = allocator.fulfillOrder(SPLIT_ORDER);
        // SPLIT_CASE ASSERTION
        Assert.assertEquals(SPLIT_CASE, actualSplitCase);
    }
}


/**
 * Run successfully, Junit test cases got passed.
 *
 * Console should print the following contents:
 *
 * ----- owd warehouse -----
 *
 * { name: owd, inventory: { banana: 6, apple: 5, peach: 8, guava: 5 } }
 *
 * ----- dm warehouse -----
 *
 * { name: dm, inventory: { banana: 6, apple: 5, pineapple: 7, peach: 2, guava: 5 } }
 *
 * ----- HAPPY_ORDER -----
 *
 * { apple: 5, pineapple: 3 }
 *
 * ----- HAPPY_CASE -----
 *
 * [ dm: { pineapple: 3 }, owd: { apple: 5 } ]
 *
 * ----- SAD_ORDER -----
 *
 * { peach: 12 }
 *
 * ----- SAD_CASE -----
 *
 * [  ]
 *
 * ----- SPLIT_ORDER -----
 *
 * { banana: 7, guava: 10 }
 *
 * ----- SPLIT_CASE -----
 *
 * [ dm: { banana: 1, guava: 5 }, owd: { banana: 6, guava: 5 } ]
 */
